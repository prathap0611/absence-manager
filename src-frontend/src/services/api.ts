export type AbsenteeStatus = 'Requested' | 'Confirmed' | 'Rejected';
export type AbsenteeType = 'vacation' | 'sickness';

export interface Absentee {
    id: number;
    userName: string;
    type: AbsenteeType;
    startDate: string | null;
    endDate: string | null;
    memberNote: string;
    status: AbsenteeStatus;
    admitterNote: string;
}

export interface PaginatedAbsentees {
    totalNumberOfPages: number;
    totalRecords: number;
    offset: number;
    limit: number;
    results: Absentee[];
}

export interface FetchAbsenteesConfig {
    currentPage: number;
}

export async function fetchAbsentees({
    currentPage,
}: FetchAbsenteesConfig): Promise<PaginatedAbsentees> {
    const limit = 10;
    const offset = currentPage * limit;
    try {
        const response = await fetch(
            '/api/absentees?' +
                new URLSearchParams({
                    offset: String(offset),
                    limit: String(limit),
                })
        );
        const absentees: PaginatedAbsentees = await response.json();
        return absentees;
    } catch (err) {
        console.error(err);
        throw new Error('Failed to fetch data');
    }
}
