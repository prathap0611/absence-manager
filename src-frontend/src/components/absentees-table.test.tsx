import { render, fireEvent } from '@testing-library/react';
import { PaginatedAbsentees } from '../services/api';
import AbsenteesTable from './absentees-table';

describe('filter tests', () => {
    const setTypeFilter = jest.fn();
    const setDateFilter = jest.fn();
    const setCurrentPage = jest.fn();

    it('should render the page correctly', () => {
        const absentees: PaginatedAbsentees = {
            totalNumberOfPages: 5,
            totalRecords: 42,
            offset: 0,
            limit: 10,
            results: [
                {
                    id: 1,
                    userName: 'user 1',
                    type: 'sickness',
                    startDate: '05/12/2020',
                    endDate: '05/12/2020',
                    memberNote: '',
                    status: 'Confirmed',
                    admitterNote: '',
                },
                {
                    id: 2,
                    userName: 'user 2',
                    type: 'vacation',
                    startDate: '05/12/2020',
                    endDate: '05/12/2020',
                    memberNote: '',
                    status: 'Rejected',
                    admitterNote: '',
                },
            ],
        };
        const { asFragment } = render(
            <AbsenteesTable
                absentees={absentees}
                typeFilter="vacation"
                setTypeFilter={setTypeFilter}
                dateFilter="03/11/2020"
                setDateFilter={setDateFilter}
                setCurrentPage={setCurrentPage}
                isLoading={false}
                currentPage={0}
            />
        );
        expect(asFragment()).toMatchSnapshot();
    });
});
