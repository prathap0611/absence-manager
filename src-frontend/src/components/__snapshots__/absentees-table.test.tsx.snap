// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filter tests should render the page correctly 1`] = `
<DocumentFragment>
  <section
    class="table-container"
  >
    <section
      class="top-container"
    >
      <section
        class="filter-container"
      >
        <label
          for="type-filter"
        >
          Choose filter type: 
          <select
            class="type-filter"
            data-testid="type-filter"
            id="type-filter"
          >
            <option
              value=""
            >
              Filter by type
            </option>
            <option
              value="vacation"
            >
              Vacation
            </option>
            <option
              value="sickness"
            >
              Sickness
            </option>
          </select>
        </label>
        <label
          for="date-filter"
        >
          Choose filter date: 
          <input
            class="date-filter"
            data-testid="date-filter"
            id="date-filter"
            type="date"
            value="03/11/2020"
          />
        </label>
      </section>
      <div
        class="absentees-count"
      >
        Total Absentees 42
      </div>
    </section>
    <table>
      <thead>
        <tr>
          <th
            scope="col"
          >
            Name
          </th>
          <th
            scope="col"
          >
            Type
          </th>
          <th
            scope="col"
          >
            Start Date
          </th>
          <th
            scope="col"
          >
            End Date
          </th>
          <th
            scope="col"
          >
            Status
          </th>
          <th
            scope="col"
          >
            Member Note
          </th>
          <th
            scope="col"
          >
            Admitter Note
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            user 1
          </td>
          <td>
            sickness
          </td>
          <td>
            12/05/2020
          </td>
          <td>
            12/05/2020
          </td>
          <td>
            Confirmed
          </td>
          <td />
          <td />
        </tr>
        <tr>
          <td>
            user 2
          </td>
          <td>
            vacation
          </td>
          <td>
            12/05/2020
          </td>
          <td>
            12/05/2020
          </td>
          <td>
            Rejected
          </td>
          <td />
          <td />
        </tr>
      </tbody>
      <tfoot>
        <tr>
          <td
            colspan="7"
          >
            <section
              class="pagination-group"
            >
              <button
                aria-label="previous page"
                class="page-btn"
                data-testid="prev-page"
                disabled=""
              >
                ◀
              </button>
              <span
                class="page-info"
              >
                1 of 5
              </span>
              <button
                aria-label="next page"
                class="page-btn"
                data-testid="next-page"
              >
                ▶
              </button>
            </section>
          </td>
        </tr>
      </tfoot>
    </table>
  </section>
</DocumentFragment>
`;
